formsfeeder.client/src/main/java/formsfeeder/client/FormsFeederClient.java

```diff
@@ -39,23 +39,29 @@
 	private final Logger baseLogger = LoggerFactory.getLogger(this.getClass());
 	
 	private final WebTarget target;
-	private final Supplier<String> correlationIdFn;
 	private final String pluginName;
-	private final Supplier<Map.Entry<String,String>> apikeyAuthenticationFn;
+	private final Map<String,Supplier<String>> headerMap;
 	String returnedCorrelationId = null;
 	
-	private FormsFeederClient(WebTarget target, Supplier<String> correlationIdFn, String pluginName,
-                              Supplier<Map.Entry<String,String>> apikeyAuthenticationFn) {
+	private FormsFeederClient(WebTarget target, String pluginName,
+							  Map<String, Supplier<String>> headerMap) {
 		this.target = target;
-		this.correlationIdFn = correlationIdFn;
 		this.pluginName = pluginName;
-		this.apikeyAuthenticationFn = apikeyAuthenticationFn;
+		this.headerMap = headerMap;
 	}
 
 	@Override
 	public DataSourceList accept(DataSourceList dataSources) throws FormsFeederClientException {
 		try {
-			String correlationIdSent = correlationIdFn != null ? CorrelationId.generate(correlationIdFn.get()) : CorrelationId.generate();
+            String correlationIdSent;
+            if(headerMap.containsKey(CorrelationId.CORRELATION_ID_HDR)) {
+                correlationIdSent = CorrelationId.generate(headerMap.get(CorrelationId.CORRELATION_ID_HDR).get());
+                headerMap.replace(CorrelationId.CORRELATION_ID_HDR, ()-> correlationIdSent);
+            } else {
+                correlationIdSent = CorrelationId.generate();
+                headerMap.put(CorrelationId.CORRELATION_ID_HDR, ()-> correlationIdSent);
+            }
+

```

Comment:
This section code concerns me a little.  Since accept() may be called multiple times and the correlationId function might provide different correlation ids for each invocation, I would probably not want to call it once and cache the result as is done here. 

I was about to provide a bunch of snippets of code changes to correct this issue when I realized none of them would work.  The problem is correlationId is _special_.  We need to evaluate it early (so that it can appear in all the log entries) whereas the rest of the headers don't need to be evaluated until much later in the code.  Trying to treat it as _not special_ just causes more headaches.

I'm thinking the idea of rolling correlationId in with the rest of the headers may be a non-starter.  Would you be OK with rolling the correlationId related code back to the way it was before while keeping the rest of the headers code  like you have here?  Sorry for the back and forth on this.

Reply:
Yep no problem. I've rolled back the correlationId stuff and added a couple of tests involving the headers being added and the custom context root.